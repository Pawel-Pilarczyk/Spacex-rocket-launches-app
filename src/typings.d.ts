declare module '*.svg' {
  const content: React.FunctionComponent<
    React.SVGAttributes<SVGElement> & {testID?: string}
  >;

  export default content;
}
