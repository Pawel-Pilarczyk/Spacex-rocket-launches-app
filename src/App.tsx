import React from 'react';
import {ApolloProvider, ApolloClient, InMemoryCache} from '@apollo/client';
import RootNavigation from './navigation';

export const API = 'https://api.spacex.land/graphql';

const client = new ApolloClient({
  uri: API,
  cache: new InMemoryCache(),
});

const App = () => {
  return (
    <ApolloProvider client={client}>
      <RootNavigation />
    </ApolloProvider>
  );
};

export default App;
