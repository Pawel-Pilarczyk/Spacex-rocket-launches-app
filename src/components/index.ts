export {default as Button} from './Button';
export {default as Typography} from './Typography';
export {default as Input} from './Input';
export {default as Item} from './Item';
export {default as ModalLoader} from './Modals/ModalLoader';
