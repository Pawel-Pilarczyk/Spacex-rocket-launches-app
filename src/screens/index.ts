export {default as Home} from './Home';
export {default as Welcome} from './Welcome';
export {default as LaunchDetails} from './LaunchDetails';
export {default as ErrorScreen} from './Error';
