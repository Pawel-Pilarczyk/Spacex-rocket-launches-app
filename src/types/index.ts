export type TNavigation = {
  navigate: (route: string, params?: any) => void;
  goBack: () => void;
};

export type TRoute = {
  params: any;
};

export type TScreenProps = {
  navigation: TNavigation;
  route?: TRoute;
};

export type TLaunch = {
  mission_name: string;
  launch_site: {
    site_name_long: string;
  };
  links: {
    article_link: string;
    video_link: string;
  };
  rocket: {
    rocket_name: string;
  };
  launch_date_utc: string;
  launch_success: boolean;
  id: string;
};

export type TLaunchesData = {
  launchesPast: Array<TLaunch>;
};
