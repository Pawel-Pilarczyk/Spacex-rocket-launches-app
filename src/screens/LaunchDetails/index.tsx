import {StyleSheet, View, Linking, ScrollView} from 'react-native';
import React from 'react';
import {TLaunch, TScreenProps} from 'src/types';
import {COLORS} from 'src/constants/colors';
import {scaling} from 'src/styles/scaling';
import {Button, Typography} from 'src/components';
import {RocketFailure, RocketSuccess} from 'src/assets/svg';

export type TParams = {
  launchData: TLaunch;
};

const ModalLaunchData = ({route}: TScreenProps) => {
  const {launchData}: TParams = route?.params || {};
  const {
    launch_date_utc,
    launch_site,
    launch_success,
    links,
    mission_name,
    rocket,
  } = launchData;
  const {article_link, video_link} = links;

  const handleOpenArticleLink = () => Linking.openURL(article_link);
  const handleOpenVideoLink = () => Linking.openURL(video_link);
  return (
    <View style={styles.wrapper}>
      <ScrollView>
        <Typography type="bold" size="22">
          Mission name
        </Typography>
        <Typography size="22">{mission_name}</Typography>
        <View style={styles.spacer} />
        <View style={styles.missionStatusWrapper}>
          <View>
            <Typography type="bold" size="22">
              Mission status
            </Typography>
            <Typography
              size="22"
              color={launch_success ? COLORS.SUCCESS : COLORS.ERROR}
              type="bold">
              {launch_success ? 'Success' : 'Failure'}
            </Typography>
          </View>
          {launch_success ? <RocketSuccess /> : null}
          {!launch_success ? <RocketFailure /> : null}
        </View>
        <View style={styles.spacer} />
        <Typography type="bold" size="22">
          Mission date
        </Typography>
        <Typography size="22">{launch_date_utc}</Typography>
        <View style={styles.spacer} />
        <Typography type="bold" size="22">
          Rocket Name
        </Typography>
        <Typography size="22">{rocket.rocket_name}</Typography>
        <View style={styles.spacer} />
        <Typography type="bold" size="22">
          Launch site
        </Typography>
        <Typography size="22">{launch_site.site_name_long}</Typography>
        <View style={styles.spacer} />
        <View style={styles.buttonsWrapper}>
          {article_link ? (
            <Button
              onPress={handleOpenArticleLink}
              style={styles.button}
              testID="articleButton">
              Open Article
            </Button>
          ) : null}
          {video_link && (
            <Button
              onPress={handleOpenVideoLink}
              type="outline"
              testID="videoButton">
              Open Video
            </Button>
          )}
        </View>
      </ScrollView>
    </View>
  );
};

export default ModalLaunchData;

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    paddingHorizontal: scaling.hs(10),
    paddingVertical: scaling.vs(10),
    backgroundColor: COLORS.WHITE,
  },
  missionStatusWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  spacer: {
    height: scaling.vs(25),
  },
  buttonsWrapper: {
    marginTop: 'auto',
    flexDirection: 'row',
  },
  button: {
    marginRight: scaling.hs(20),
  },
});
