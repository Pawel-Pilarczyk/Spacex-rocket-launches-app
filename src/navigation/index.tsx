import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import {ROUTES} from 'src/constants/routes';
import {Home, Welcome, LaunchDetails, ErrorScreen} from 'src/screens';

const RootStack = createStackNavigator();

const RootNavigation = () => {
  return (
    <NavigationContainer>
      <RootStack.Navigator
        screenOptions={{
          headerTitleAlign: 'center',
          headerBackTitle: '',
        }}>
        <RootStack.Screen
          name={ROUTES.WELCOME}
          component={Welcome}
          options={{headerShown: false}}
        />
        <RootStack.Screen
          name={ROUTES.HOME}
          component={Home}
          options={{headerTitle: 'Missions List'}}
        />
        <RootStack.Screen
          name={ROUTES.DETAILS}
          component={LaunchDetails}
          options={{headerTitle: 'Mission Details'}}
        />
        <RootStack.Screen name={ROUTES.ERROR} component={ErrorScreen} />
      </RootStack.Navigator>
    </NavigationContainer>
  );
};

export default RootNavigation;
