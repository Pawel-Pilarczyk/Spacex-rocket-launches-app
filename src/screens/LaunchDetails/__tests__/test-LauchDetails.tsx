/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * @format
 */

import 'react-native';
import React from 'react';
import LaunchDetails from 'src/screens/LaunchDetails';
import {TNavigation, TLaunch, TRoute} from 'src/types';

// Note: test renderer must be required after react-native.
import {render} from '@testing-library/react-native';

const launchData: TLaunch = {
  id: '1',
  launch_date_utc: 'Thu Mar 31 2022 14:56:49 GMT+0200',
  launch_site: {site_name_long: 'TEST SIDE'},
  launch_success: true,
  links: {
    article_link: '',
    video_link: '',
  },
  mission_name: 'TEST MISSION',
  rocket: {
    rocket_name: 'TEST ROCKET',
  },
};

const videoLaunchData: TLaunch = {
  ...launchData,
  links: {
    article_link: '',
    video_link: 'Link',
  },
};
const articleLaunchData: TLaunch = {
  ...launchData,
  links: {
    article_link: 'Link',
    video_link: '',
  },
};
const articleAndVideoLaunchData: TLaunch = {
  ...launchData,
  links: {
    article_link: 'Link',
    video_link: 'Link',
  },
};

const navigation: TNavigation = {
  navigate: (route: string, params: any) => {},
  goBack: () => {},
};

it('should render correctly', () => {
  const route: TRoute = {
    params: {
      launchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  ).toJSON();
  expect(launchDetails).toMatchSnapshot();
});

it('should render Open Video Button and should not render Article button', () => {
  const route: TRoute = {
    params: {
      launchData: videoLaunchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  );

  expect(launchDetails.queryAllByTestId('videoButton').length).toEqual(1);
  expect(launchDetails.queryAllByTestId('articleButton').length).toEqual(0);
});

it('should render Open Article Button and should not render Video button', () => {
  const route: TRoute = {
    params: {
      launchData: articleLaunchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  );

  expect(launchDetails.queryAllByTestId('videoButton').length).toEqual(0);
  expect(launchDetails.queryAllByTestId('articleButton').length).toEqual(1);
});

it('should render Open Article Button and Video button', () => {
  const route: TRoute = {
    params: {
      launchData: articleAndVideoLaunchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  );

  expect(launchDetails.queryAllByTestId('videoButton').length).toEqual(1);
  expect(launchDetails.queryAllByTestId('articleButton').length).toEqual(1);
});

it('should not render Open Article Button and Video button', () => {
  const route: TRoute = {
    params: {
      launchData: launchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  );

  expect(launchDetails.queryAllByTestId('videoButton').length).toEqual(0);
  expect(launchDetails.queryAllByTestId('articleButton').length).toEqual(0);
});

it('should display Lauch Data', () => {
  const route: TRoute = {
    params: {
      launchData: launchData,
    },
  };
  const launchDetails = render(
    <LaunchDetails navigation={navigation} route={route} />,
  );

  expect(launchDetails.queryAllByTestId('videoButton').length).toEqual(0);
  expect(launchDetails.queryAllByTestId('articleButton').length).toEqual(0);
});
