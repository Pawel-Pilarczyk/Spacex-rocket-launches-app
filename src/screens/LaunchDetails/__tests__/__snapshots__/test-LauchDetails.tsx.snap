// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 16,
    }
  }
>
  <RCTScrollView>
    <View>
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "800",
            },
          ]
        }
      >
        Mission name
      </Text>
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "400",
            },
          ]
        }
      >
        TEST MISSION
      </Text>
      <View
        style={
          Object {
            "height": 41,
          }
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": "#000",
                  "fontSize": 44,
                  "fontWeight": "800",
                },
              ]
            }
          >
            Mission status
          </Text>
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": "#22C55E",
                  "fontSize": 44,
                  "fontWeight": "800",
                },
              ]
            }
          >
            Success
          </Text>
        </View>
        <starting-rocket.svg />
      </View>
      <View
        style={
          Object {
            "height": 41,
          }
        }
      />
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "800",
            },
          ]
        }
      >
        Mission date
      </Text>
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "400",
            },
          ]
        }
      >
        Thu Mar 31 2022 14:56:49 GMT+0200
      </Text>
      <View
        style={
          Object {
            "height": 41,
          }
        }
      />
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "800",
            },
          ]
        }
      >
        Rocket Name
      </Text>
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "400",
            },
          ]
        }
      >
        TEST ROCKET
      </Text>
      <View
        style={
          Object {
            "height": 41,
          }
        }
      />
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "800",
            },
          ]
        }
      >
        Launch site
      </Text>
      <Text
        style={
          Array [
            undefined,
            Object {
              "color": "#000",
              "fontSize": 44,
              "fontWeight": "400",
            },
          ]
        }
      >
        TEST SIDE
      </Text>
      <View
        style={
          Object {
            "height": 41,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": "auto",
          }
        }
      >
        
      </View>
    </View>
  </RCTScrollView>
</View>
`;
